### DISCUSSION TITLE
Duplicate detection -- data importer

### DISCUSSION THREAD

**Discussion by bartekd123** [firefly-iii/firefly-iii#6633] [General]
Sometimes I have 2 transactions that look exactly the same, the same amount and the same date. Is there a way of not importing duplicates, but if there is a duplicate of 2 in the same file on the same date, that it imports both?

My situation, I import all the transactions from the current month using data importer. but there will be some transactions that look exactly the same. It imports the first one, but then the 2nd one it skips because that transaction already exists.

My question is if there is a way of doing the duplicate check before the file starts to run? That way if there are 2 transactions in the same file, it would flag the 2nd transaction being a duplicate.

**Comment by JC5** [comment]
It will flag the second one, but only during the import. It will not warn you before it starts, there is no way to do that from the importer.

**Reply by bartekd123** [reply]
Thanks James. What did you mean it will flag it, but by "only during the import"? Not sure if I am missing something. Or do I just need to find those transactions and duplicate them?

Could it be a feature request to do this? When it is doing the import, if there are duplicates in the same file, then import both transactions, otherwise do the duplicate detection?

**Reply by JC5** [reply]
I'm not really sure what you expect the tool to do. You can do two things, see the example file below:

```
description,date,amount,source,destination,ignored
Transaction A,2022-11-12,12.50,Checking Account,Amazon,ignore_me_a
Transaction A,2022-11-12,12.50,Checking Account,Amazon,ignore_me_a
Transaction A,2022-11-12,12.50,Checking Account,Amazon,ignore_me_b
```

1. You can make sure that the importer ignores the second transaction in the file (exactly the same line as the first). It will not even attempt to import it, it will be filtered out. See below:

```
[2022-11-12 06:09:46] local.WARNING: Going to skip line #1 because it's in the file twice. This may reset the count below.
```

2. There is also the "normal" duplicate detection that is handled during the import itself. This is (technically speaking) done by Firefly III itself. It will ignore the third line because the second line was imported successfully.

```
local.ERROR: Submission error: 0 ["Duplicate of transaction #226."]
local.DEBUG: This is a duplicate transaction error
local.DEBUG: Add error on index #1 (line no. 2): transactions.0.description: Duplicate of transaction #226. (original value: "Transaction A")
```

**Reply by bartekd123** [reply]
The problem is, I DO want it to import the duplicates sometimes. Below is an example, and I'll use Amazon.

 - Yesterday I import all the transactions for November, so I am caught up to November 10th in firefly
 - After I do the import (yesterday the 11th) I buy some socks on amazon, cost is $2.22
 - An hour later I think, hey I could use another pair of the same socks. Buy it again (same price, same credit card, same everything)
 - today comes along, I do another import. This includes all of November

What happens
 - It imports all the transactions from the 11th (since all the ones from the 10th and before were already there and duplication prevented them to be imported again)
 - There is only one transaction for the amazon order, even though I bought it twice (since it was seen as a duplicate)

What I would like to see happen
 - If there is a new transaction being added, if there is duplicates within the same csv file, it adds them both. If it is a duplicate of one that is already in firefly (before I started my import), then it flags it as a duplicate and skips it

**Reply by JC5** [reply]
That can be done by unselecting "filter duplicate lines" and enabling the import duplicate check. Your milage may vary though, if there is nothing to tell the transactions apart (like a transaction ID) this may fail.

**Reply by bartekd123** [reply]
There is nothing distinguishing the lines in the source CSV file, I just looked at an example transaction that I had this issue with yesterday. 


I just tried setting this in the json, and it only imports one of them, skipping the 2nd one:
    "ignore_duplicate_lines": false,
    "ignore_duplicate_transactions": true,

**Reply by bartekd123** [reply]
Here is what my CSV looks like:

```
Date,Transaction,Name,Memo,Amount
11/8/2022,OTHER,EFT Withdrawal to PAYBRIGHT,To PAYBRIGHT,-10.10
11/8/2022,OTHER,EFT Withdrawal to PAYBRIGHT,To PAYBRIGHT,-10.10

```

**Reply by JC5** [reply]
Ah, I see what you mean. I was able to import according to your preferences using these settings:

```
{
    "duplicate_detection_method": "classic",
    "ignore_duplicate_lines": false,
    "ignore_duplicate_transactions": false
}
```

`ignore_duplicate_transactions` is the old version of the duplicate detection settings and when you set it to `true`, it will actually set `duplicate_detection_method` to `classic` so I left it on `false`. You can also remove it entirely.

The effect of these setting is:

- Both transactions in the CSV file are submitted to Firefly III.
- Both will be checked for being present already in Firefly III.
- Best case scenario: only the first line will be imported.
- Worst case: neither will be imported (if it existed already)

**Reply by bartekd123** [reply]
Thanks, I already had "duplicate_detection_method": "classic" set in my json file. 

What I wanted was to have both transactions imported. Since they didn't exist in firefly before my import. They are both legit transactions, which just look the same. I guess there is no way to make it import both transactions, but if there are any duplicates that are already in my firefly from before the import, it will treat them as duplicates?

**Reply by JC5** [reply]
No, that option is not available unfortunately.

**Reply by bartekd123** [reply]
Ok thanks. Could I create a feature request for this, and it can be considered down the line at some point? Or do you think its not worth it?

**Reply by JC5** [reply]
I'm not inclined to do so, to be honest. It's a pretty niche request, and I'm afraid it will make all of the (already confusing) duplicate detection methods even more confusing.

What could help in your case is adding a column to your CSV files with an identifier in them. This column should be imported as "external id". If it is based on the line content and the date its uploaded, it should help doing what you want.

**Reply by bartekd123** [reply]
Thats fair, thanks for your replies. 
Adding an external ID wouldn't really work for me as I would have to take all the existing ones I uploaded previously and cross reference the ID's I had for those transactions. 

Its all good, Ill just periodically monitor and compare my balance in firefly and bank balances, and see if there are any discrepancies with the balance and see if its things like this happening.
